From 89f926065d911296213c39b71820ecc212e77f18 Mon Sep 17 00:00:00 2001
From: Tushar Darote <quic_tdarote@quicinc.com>
Date: Thu, 9 Oct 2025 22:42:17 +0200
Subject: [PATCH 3/9] feat(tflite): Add dynamic OpenCL library loading support

Use pkg-config to detect OpenCL installation and set library name based on
OpenCL version. Pass library name as compile-time definition to OpenCL wrapper
and modify wrapper to use configured name instead of hardcoded defaults.
This fixes OpenCL library loading issues on systems with non-standard
installations and various Linux distributions.

Upstream-Status: Submitted [https://github.com/tensorflow/tensorflow/pull/109377]
Signed-off-by: Tushar Darote <quic_tdarote@quicinc.com>
---
 tensorflow/lite/CMakeLists.txt                     | 14 ++++++++++++++
 tensorflow/lite/delegates/gpu/cl/opencl_wrapper.cc |  2 +-
 2 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/tensorflow/lite/CMakeLists.txt b/tensorflow/lite/CMakeLists.txt
index cdd38ff2086..2971047b3fc 100644
--- a/tensorflow/lite/CMakeLists.txt
+++ b/tensorflow/lite/CMakeLists.txt
@@ -746,6 +746,20 @@ if (NOT BUILD_SHARED_LIBS)
   list(APPEND TFLITE_TARGET_PUBLIC_OPTIONS "-DTFL_STATIC_LIBRARY_BUILD")
 endif()
 
+find_package(PkgConfig)
+pkg_check_modules(OPENCL opencl)
+
+if(OPENCL_FOUND)
+  string(REGEX MATCH "^[0-9]+" OPENCL_VERSION_MAJOR ${OPENCL_VERSION})
+  set(CL_LIB_NAME "libOpenCL.so.${OPENCL_VERSION_MAJOR}")
+else()
+  set(CL_LIB_NAME "libOpenCL.so")
+endif()
+
+target_compile_definitions(tensorflow-lite
+  PRIVATE CL_LIB_NAME=\"${CL_LIB_NAME}\"
+)
+
 target_compile_options(tensorflow-lite
   PUBLIC ${TFLITE_TARGET_PUBLIC_OPTIONS}
   PRIVATE ${TFLITE_TARGET_PRIVATE_OPTIONS}
diff --git a/tensorflow/lite/delegates/gpu/cl/opencl_wrapper.cc b/tensorflow/lite/delegates/gpu/cl/opencl_wrapper.cc
index 2eb95df35ae..b8229ec1f96 100644
--- a/tensorflow/lite/delegates/gpu/cl/opencl_wrapper.cc
+++ b/tensorflow/lite/delegates/gpu/cl/opencl_wrapper.cc
@@ -119,7 +119,7 @@ absl::Status LoadOpenCLOnce() {
   static const char* kClLibName =
       "/System/Library/Frameworks/OpenCL.framework/OpenCL";
 #else
-  static const char* kClLibName = "libOpenCL.so";
+  static const char* kClLibName = CL_LIB_NAME;
 #endif
 #ifdef __ANDROID__
   libopencl = AndroidDlopenSphalLibrary(kClLibName, RTLD_NOW | RTLD_LOCAL);
-- 
2.34.1

